src/samplereader/FragmentedSampleReader.cpp

```diff
@@ -0,0 +1,462 @@
+      m_encrypted.SetData(m_sampleData.GetData(), m_sampleData.GetDataSize());
+    else if (decrypterPresent && m_decrypter == nullptr && !useDecryptingDecoder)
+      m_sampleData.SetData(m_encrypted.GetData(), m_encrypted.GetDataSize());
+
+    if (m_decrypter)
+    {
+      // Make sure that the decrypter is NOT allocating memory!
+      // If decrypter and addon are compiled with different DEBUG / RELEASE
+      // options freeing HEAP memory will fail.
+      m_sampleData.Reserve(m_encrypted.GetDataSize() + 4096);
+      if (AP4_FAILED(result =
+                         m_decrypter->DecryptSampleData(m_poolId, m_encrypted, m_sampleData, NULL)))
+      {
+        LOG::Log(LOGERROR, "Decrypt Sample returns failure!");
+        if (++m_failCount > 50)
+        {
+          Reset(true);
+          return result;
+        }
+        else
+        {
+          m_sampleData.SetDataSize(0);
+        }
+      }
+      else
+      {
+        m_failCount = 0;
+      }
+    }
+    else if (useDecryptingDecoder)
+    {
+      m_sampleData.Reserve(m_encrypted.GetDataSize() + 1024);
+      m_singleSampleDecryptor->DecryptSampleData(m_poolId, m_encrypted, m_sampleData, nullptr, 0,
+                                                 nullptr, nullptr);
+    }
+
+    if (m_codecHandler->Transform(m_sample.GetDts(), m_sample.GetDuration(), m_sampleData,
+                                  m_track->GetMediaTimeScale()))
+    {
+      m_codecHandler->ReadNextSample(m_sample, m_sampleData);
+    }
+  }
+
+  m_dts = (m_sample.GetDts() * m_timeBaseExt) / m_timeBaseInt;
+  m_pts = (m_sample.GetCts() * m_timeBaseExt) / m_timeBaseInt;
+
+  m_codecHandler->UpdatePPSId(m_sampleData);
+
+  return AP4_SUCCESS;
+}
+
+void FragmentedSampleReader::Reset(bool bEOS)
+{
+  AP4_LinearReader::Reset();
+  m_eos = bEOS;
+  if (m_codecHandler)
+    m_codecHandler->Reset();
+}
+
+uint64_t FragmentedSampleReader::GetDuration() const
+{
+  return (m_sample.GetDuration() * m_timeBaseExt) / m_timeBaseInt;
+}
+
+bool FragmentedSampleReader::IsEncrypted() const
+{
+  return (m_decrypterCaps.flags & SSD::SSD_DECRYPTER::SSD_CAPS::SSD_SECURE_PATH) != 0 &&
+         m_decrypter != nullptr;
+}
+
+bool FragmentedSampleReader::GetInformation(kodi::addon::InputstreamInfo& info)
+{
+  if (!m_codecHandler)
+    return false;
+
+  bool edChanged(false);
+  if (m_bSampleDescChanged && m_codecHandler->m_extraData.GetDataSize() &&
+      !info.CompareExtraData(m_codecHandler->m_extraData.GetData(),
+                             m_codecHandler->m_extraData.GetDataSize()))
+  {
+    info.SetExtraData(m_codecHandler->m_extraData.GetData(),
+                      m_codecHandler->m_extraData.GetDataSize());
+    edChanged = true;
+  }
+
+  AP4_SampleDescription* desc(m_track->GetSampleDescription(0));
+  if (desc->GetType() == AP4_SampleDescription::TYPE_MPEG)
+  {
+    switch (static_cast<AP4_MpegSampleDescription*>(desc)->GetObjectTypeId())
```

Comment:
I understand this is a 'safe' way to go about this, but I would have thought that since we already checked the type of sampledescription we're dealing with (`if (desc->GetType() == AP4_SampleDescription::TYPE_MPEG)`) then the static cast should be completely safe? If there was no `GetType` method then yes I could see why we would do it this way.

Reply:
ok